### PR TITLE
New app setup branch

### PR DESCRIPTION
**Pull request by alyssakelley**
I made a different branch before this one but I do not think the parent was the master so I re-did it. Hopefully this one is now accurate, sorry about that.

Currently this file creates the yaml file based on the CL input, and also creates the Namespace according to that file. Next steps are to add the team users, create the bucket and github repo. Also included the unittests that I currently have.

### CODE REVIEW

File: new_app_setup.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+        yaml.dump(yml_file_kubernetes_data, outfile, default_flow_style=False)
+
+    print("The YAML file for {} pods were created and was "
+          "saved as {}".format(team_name, pods_filename))
+
+    return yml_file_kubernetes_data
+
+
+def create_kubernetes(filename, team_name, type_of_create):
+    subprocess.call(['kubectl', 'create', '-f', filename])
```
Comment (slively) [inline new_app_setup.py]: check_call

File: new_app_setup.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    # print("Bucket {} was created.".format(bucket.name))
+
+
+def get_congif_yml_file(team_name):
+    config_filename = team_name+'_config_file.yml'
+
+    terminal_command = ['kubectl', 'config', 'view']
+
+    with open(config_filename, 'w') as outfile:
+        subprocess.call(terminal_command, stdout=outfile)
```
Comment (slively) [inline new_app_setup.py]: check_call

File: requirements.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+PyYAML
```
Comment (slively) [inline requirements.txt:1]: this should include a version number `PyYAML=1.2.3`

**Review by slively — CHANGES_REQUESTED** [review]

File: new_app_setup.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    with open(namespace_filename, 'w') as outfile:
+        yaml.dump(yml_file_kubernetes_data, outfile, default_flow_style=False)
+
+    print("The YAML file for {} was created and was "
+          "saved as {}".format(team_name, namespace_filename))
+
+    return yml_file_kubernetes_data
+
+
+def create_kubernetes_pods_yml(pods_filename, team_name, ram_requested,
```
Comment (slively) [inline new_app_setup.py]: we don't need to create a pod for them, just a namespace

File: new_app_setup.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+def send_config_email_to_users(team_name, email_to, config_filename):
+    '''
+    Note: This code is going to change to reflect which email address we
+    would be sending this from. This successfully sends emails to GMAIL
+    and UOREGON accounts, I have not tried to send to any
+    other emails.
+    '''
+    # QUESTION: What email address do you want me to have these
+    # config files send from?
+    MVP_email_address = 'abkelley97@gmail.com'
```
Comment (slively) [inline new_app_setup.py]: mvpstudiooregon@gmail.com for now would work

File: new_app_setup.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+        server.sendmail(MVP_email_address, receiver_email_address, text)
+        print ("The config file for {} has successfully been sent "
+               "to {}".format(team_name, email_to))
+
+    server.close()
+
+
+def add_users(filename, group_user, list_of_users):
+    '''
+    QUESTIONS: I am still unclear after looking into this a good amount how we
```
Comment (slively) [inline new_app_setup.py]: You are on the right track. We can start by just creating a single ServiceAccount for the team. Then a Role within their namespace that has full access to every thing (you can use a *), then a RoleBinding that applies that new Role to the new ServiceAccount. Note, we don't want to use ClusterRole or ClusterRoleBinding as that would give them access to everything. If it helps we can go through doing this manually on Tuesday.

Example of a Role: https://kubernetes.io/docs/reference/access-authn-authz/rbac/#role-and-clusterrole
Example of a RoleBinding: https://kubernetes.io/docs/reference/access-authn-authz/rbac/#rolebinding-and-clusterrolebinding

File: new_app_setup.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+def create_kubernetes(filename, team_name, type_of_create):
+    subprocess.call(['kubectl', 'create', '-f', filename])
+
+    print("The Kubernetes {} for {} was "
+          "created.".format(type_of_create, team_name))
+
+
+def create_bucket(team_name):
+    pass
+    # export PROJECT_ID=$(gcloud config get-value team_name)
```
Comment (slively) [inline new_app_setup.py]: our project_id is mvpstudio this could be read in through a command line argument

Comment (slively) [reply new_app_setup.py]: We want to create 2 buckets, one for file storage and another for database backups. Something like, 
`statg-projectname-files` & `stage-projectname-db-backups`

File: new_app_setup.py (hunk 5 above)
Comment (slively) [reply new_app_setup.py]: could also be a command line arg

Comment (slively) [reply new_app_setup.py]: we could also look into using the Gmail API for this

File: new_app_setup.py (hunk 4 above)
Comment (alyssakelley) [reply new_app_setup.py]: So then we never set the CPU and RAM for the team? I thought this was a main thing that I was supposed to do for this script, is it not?

File: new_app_setup.py (hunk 5 above)
Comment (alyssakelley) [reply new_app_setup.py]: Ok I can put that email address in, can you message me the password for this account so I can continue to test it?

File: new_app_setup.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+    base_path = os.path.abspath(os.path.dirname(__file__))
+
+    parser = argparse.ArgumentParser(description='Usage: python3 '
+                                     'new_app_setup.py <Team name> '
+                                     '<List of users>')
+
+    parser.add_argument('team_name', type=str,
+                        help='Team name')
+
+    parser.add_argument('group_user', type=str, help='An email address that '
```
Comment (oliverdain) [inline new_app_setup.py]: this isn't reflected in the usage on line 23.

This is a matter of personal preference, but I generally prefer explicit arguments to multiple positional arguments. With explicit args the order doesn't matter, the command line is more clear and thus less bug prone.

For example, `new_app_setup.py my_team oliver@dains.org alyssa@uoregon.edu` -- OOPS - I just made myself the "primary" email for the team by accident. vs. `new_app_setup.py --primary=alyssa@uoregon.edu --name=my_team oliver@dains.org scott@mvpstudio.org`

File: new_app_setup.py [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+                        'will be used for the entire group as a whole.')
+
+    parser.add_argument('-LOU', '--list_of_users',
+                        required=False,
+                        help='A list of the users for this team.')
+
+    parser.add_argument('-RAM', '--RAM', type=int, required=False,
+                        help='How many gigabytes of RAM your team is '
+                        'requesting, the deault is 2G.')
```
Comment (oliverdain) [inline new_app_setup.py]: you can do `default=2` here and argparse will set that if the arg isn't given

File: new_app_setup.py [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+
+    parser.add_argument('-CPU', '--CPU', type=int, required=False,
+                        help="How much CPU your team is requesting, the "
+                        "default is one.")
```
Comment (oliverdain) [inline new_app_setup.py]: default = 1

File: new_app_setup.py [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+
+    args = parser.parse_args()
+
+    team_name = args.team_name.replace(" ", "-").replace("_", "-").lower()
+
+    namespace_filename = team_name+"_namespace.yml"
```
Comment (oliverdain) [inline new_app_setup.py]: I think pep8 requires spaces here

File: new_app_setup.py [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+    pods_filename = team_name+"_pods.yml"
+
+    create_kubernetes_namespace_yml(namespace_filename, team_name)
+
+    create_kubernetes(namespace_filename, team_name, "namespace")
+
+    if args.RAM is not None:
+        ram_requested = args.RAM
+    else:
+        ram_requested = '2G'  # default
```
Comment (oliverdain) [inline new_app_setup.py]: as above, let `argparse` handle the defaults for you

File: new_app_setup.py (hunk 4 above)
Comment (oliverdain) [reply new_app_setup.py]: Yes, one of the main things this should do is set limits. But the limits go on the namespace and are set via a `ResourceQuota`: https://kubernetes.io/docs/tasks/administer-cluster/manage-resources/quota-memory-cpu-namespace/

Once you have a resource quota set on a namespace then all pods must specify how much memory and CPU they want.

For example, let's say the namespace is "example" and we put a `ResourceQuota` on it for 2 CPUs and 2GB of RAM. Then users of that namespace:

1. Create a pod that requests 1 CPU and 1.5 GB of RAM: no problem - k8's will schedule that and they have 1CPU and 1/2 GB of RAM *remaining* in the quota for their namespace
2. Then they try to create a pod that wants 0.5 CPU and 2GB of RAM. k8's will **not** schedule this because the 2GB of RAM is more than what's remaining in their quote
3. The devs then kill the pods in (1). Now there's enough CPU and RAM so the pod from (2) will get run.

In other words: the quota is the total amount of resources that can be used by all the pods in a namespace. Pods request resources which count against that quota.

File: new_app_setup.py [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+# QUESTION: When running, I am getting the warning:
+
+# "error: error validating "test-team-name_pods.yml": error validating
+# data: ValidationError(Pod.spec.containers): invalid type for
+# io.k8s.api.core.v1.PodSpec.containers: got "map", expected "array";
+# if you choose to ignore these errors, turn validation off with
+# --validate=false
+# and I know that is because name does not have the hyphen indicating an array
+# in front of it, but I cannot seem to figure out how to do this...
+# anything I'm missing?
```
Comment (oliverdain) [inline new_app_setup.py]: it's saying "containers: got "map", expected "array" - that is containers on line 117 is supposed to be a **list** of containers but you've given it a dict. A dict in python == a map in YAML. So line 117 should be something like:

```
spec = dict(
    containers=[
        dict( name=team_name, ....)
```

Note that it's much more common in Python to just use `{}` for a dict instead of the explicit dict constructor. That is, `dict(a=1, b=2)` and `{'a': 1, 'b': 2}` create exactly the same object by the 2nd is a much more common way to see it written.

File: new_app_setup.py [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+            ),
+        spec=dict(
+                containers=dict(
+                    name=team_name,  # QUESTION HERE -> see below
+                    resources=dict(
+                        limits=dict(
+                            cpu="1",
+                            memory="2G"
+                        ),
+                        requests=dict(
```
Comment (oliverdain) [inline new_app_setup.py]: so here the pods are requesting to *use* `cpu_requested` and `ram_requested` -- pods are the apps that run. What we want to do is create the **limits** on how much all the pods can use but leave it up to the teams to decide how many pods to run, what each pod does, and how much of the available resources each pod should use.

File: new_app_setup.py [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+    with open(pods_filename, 'w') as outfile:
+        yaml.dump(yml_file_kubernetes_data, outfile, default_flow_style=False)
+
+    print("The YAML file for {} pods were created and was "
+          "saved as {}".format(team_name, pods_filename))
+
+    return yml_file_kubernetes_data
+
+
+def create_kubernetes(filename, team_name, type_of_create):
```
Comment (oliverdain) [inline new_app_setup.py]: `create_kubernetes` sounds like it creates a k8's cluster. How about `kube_apply` or something?

File: new_app_setup.py (hunk 1 above)
Comment (oliverdain) [reply new_app_setup.py]: s/create/apply/

`create` can create a resource but not change it. `apply` says "make it so" - if it doesn't exist it'll be created and if it does exist it'll be changed. Making that simple change means that if, for example, we decided to grant a team more resources we could just re-run this script with different CPU and RAM values.

File: new_app_setup.py [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+
+def create_bucket(team_name):
+    pass
+    # export PROJECT_ID=$(gcloud config get-value team_name)
+    # storage_client = storage.Client()
+    # bucket = storage_client.create_bucket(team_name)
+    # print("Bucket {} was created.".format(bucket.name))
+
+
+def get_congif_yml_file(team_name):
```
Comment (oliverdain) [inline new_app_setup.py:140]: s/congif/config/

But, as above, `get_config_yml_file` isn't very descriptive - there's lots of things you can configure.

Might also be a good idea to use triple-quote comments for any method whose purpose isn't **very** obvious to say what that method does.

File: new_app_setup.py [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+    # config files send from?
+    MVP_email_address = 'abkelley97@gmail.com'
+
+    # This line will be the actual one used to reflect the password
+    # for whatever email we send from.
+    # MVP_email_address_password = os.environ.get
+    #                               ('MVP_EMAIL_ADDRESS_PASSWORD')
+
+    # TEMP for my testing right now:
+    MVP_email_address_password = os.environ[str('PROF_EMAIL_PASS')]
```
Comment (oliverdain) [inline new_app_setup.py]: nice to get the password from the environment!

File: new_app_setup.py [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+    create_kubernetes_pods_yml(pods_filename, team_name, ram_requested,
+                               cpu_requested)
+
+    create_kubernetes(pods_filename, team_name, "pods")
+
+    config_file = get_congif_yml_file(team_name)
+
+    path_to_congif_file = os.path.join(base_path, config_file)
+
+    send_config_email_to_users(team_name, args.group_user, config_file)
```
Comment (oliverdain) [inline new_app_setup.py]: Currently I think this does the following:

1. Create a namespace .yml file
2. Create a pod .yml file
3. Create the pod in k8's (but you do not create the namespace)
4. email the team their pod config

Eventually (not all in this PR), I think we want this to:

1. Create the namespace yml and actually create the namespace
2. Create a ResourceQuota yml and apply it to that namespace
3. Create k8's users/credentials for (1) the team and (2) the CI. These contain for `kubectl` to know how to find the cluster, etc. but, importantly, they also contain the cryptographic keys that are used to authenticate to k8's so we shouldn't email these
4. Set up a CI build including adding the keys as a secret
5. Set up 2 buckets
6. Create a gcloud account for the team so they can access those buckets
7. email the users some general instructions for using things, pointers to example .yml files for things they might want to do, etc.

I think we'll give team members a of copy of the team's k8's credentials and gcloud credentials via thumb drive or something since that's the actual secret they use to authenticate.

For this PR maybe just focus on (1) and (2)

File: new_app_setup.py [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+    #                               ('MVP_EMAIL_ADDRESS_PASSWORD')
+
+    # TEMP for my testing right now:
+    MVP_email_address_password = os.environ[str('PROF_EMAIL_PASS')]
+
+    email_contents_subject = ("Kubernetes config file for "
+                              "Team " + team_name + " at Hack For A Cause")
+
+    email_contents_body = ("Hi there,\nAttached to this email is your "
+                           "Kubernetes configuration file for your team " +
```
Comment (oliverdain) [inline new_app_setup.py]: the config file they will need is the one that contains their credentials which we probably don't want to send via email

File: new_app_setup.py [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+    msg = MIMEMultipart()
+    msg['From'] = MVP_email_address
+    msg['Subject'] = email_contents_subject
+
+    msg.attach(MIMEText(email_contents_body, 'plain'))
+
+    attachment = open(config_filename, "rb")
+
+    # Allows us to upload the attachment
+    part = MIMEBase('application', 'octet-stream')
```
Comment (oliverdain) [inline new_app_setup.py]: nice!!!

File: new_app_setup.py (hunk 6 above)
Comment (oliverdain) [reply new_app_setup.py]: Service accounts: "User accounts are for humans. Service accounts are for processes, which run in pods." from https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#user-accounts-vs-service-accounts

From https://kubernetes.io/docs/reference/access-authn-authz/authentication/#users-in-kubernetes:

> All Kubernetes clusters have two categories of users: service accounts managed by Kubernetes, and normal users.

> Normal users are assumed to be managed by an outside, independent service. An admin distributing private keys, a user store like Keystone or Google Accounts, even a file with a list of usernames and passwords. In this regard, Kubernetes does not have objects which represent normal user accounts. Normal users cannot be added to a cluster through an API call.

> In contrast, service accounts are users managed by the Kubernetes API. They are bound to specific namespaces, and created automatically by the API server or manually through API calls. Service accounts are tied to a set of credentials stored as Secrets, which are mounted into pods allowing in-cluster processes to talk to the Kubernetes API.

In short - users should be added via `gcloud` as `gcloud` is the identity provider for our cluster. I think the relevant docs are here: https://cloud.google.com/kubernetes-engine/docs/how-to/iam

I **think** what we need to do is:

1. Use gcloud to create users in our organization with role `roles/container.developer`
2. Use `Role` and `RoleBinding` as @slively describes to limit the user to only their namespace

File: test_new_app_setup.py [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+import subprocess
+import argparse
+from unittest.mock import patch
+import io
+import sys
+
+base_path = os.path.abspath(os.path.dirname(__file__))
+
+path_to_test_namesapce_yml = os.path.join(base_path,
+                                'test_data/kubernetes/test_namespace1.yml')
```
Comment (oliverdain) [inline test_new_app_setup.py]: what are you doing with this test file?

**Review by oliverdain — CHANGES_REQUESTED** [review]
Sorry it took me so long to get to this PR!

**Comment by oliverdain** [comment]
https://cloud.google.com/solutions/prep-kubernetes-engine-for-prod has a good overview of how to create users and limit them to a single namespace in the "Managing identity and access" section. Note that the "service account" mentioned in those docs is a gcloud service account and **not** a k8's one which is a little confusing. In any event, I don't think we'll need a service account of either type though we could conceivably use a gcloud service account for the CI.

File: new_app_setup.py (hunk 4 above)
Comment (alyssakelley) [reply new_app_setup.py]: Thank you, your description and link were very helpful!

File: new_app_setup.py (hunk 18 above)
Comment (alyssakelley) [reply new_app_setup.py]: Ok, I will focus on those first two steps. In regards to making the namespace yml and actual namespace, is this not the case with these two lines?

    create_kubernetes_namespace_yml(namespace_filename, team_name)

    create_kubernetes(namespace_filename, team_name, "namespace")

My create_kubermetes function does the following:

    subprocess.check_call(['kubectl', 'create', '-f', filename])

and when I try "kubectl get namespaces" I see the ones that were created (which I need to go back and delete).

File: new_app_setup.py (hunk 19 above)
Comment (alyssakelley) [reply new_app_setup.py]: So should I delete all of the email information and not send anything to the team until later on when we focus on Step #7?

File: new_app_setup.py (hunk 6 above)
Comment (alyssakelley) [reply new_app_setup.py]: Alright, so I am making the user as so:

        GROUP = 'mvpstudio'
        RESOURCE_NAME = team_name
        ROLE_ID = 'roles/container.developer'


        subprocess.check_call(['gcloud', GROUP, 'add'+'-'+'iam'+'-'+'policy'+'-'+'binding',
                RESOURCE_NAME, '--member', 'user'+':'+group_user, '--role', ROLE_ID])


and then I see there are options to make a Role and RoleBinding yaml, or to just have a command line do this with specific flags. 

Right now I am working with this:

      def create_role_yml(role_filename, team_name, group_user):
         """
         https://kubernetes.io/docs/reference/access-authn-authz/rbac/#role-and-clusterrole
         """
           yml_file_kubernetes_data = dict(

             apiVersion='rbac.authorization.k8s.io/v1',
             kind='Role',
               metadata=dict(
                 namespace=team_name,
                    name=group_user
               )
               rules=dict(
                   apiGroups=[""],
                    resourses=["pods"],
                    verbs=["get", "watch", "list"]
                 )
                )

           with open(role_filename, 'w') as outfile:
              yaml.dump(yml_file_kubernetes_data, outfile, default_flow_style=False)
   
          subprocess.check_call(['kubectl', 'create', 'role', role_filename])

       def create_rolebinding_yml(rolebinding_filename, team_name, group_user):

          yml_file_kubernetes_data = dict(

              apiVersion='rbac.authorization.k8s.io/v1',
             kind='RoleBinding',
                metadata=dict(
                  name=read-pods,
                  namespace=team_name,
                    )
             subjects=dict(
                  kind= User,
                  name=group_user,
                  apiGroup=rbac.authorization.k8s.io,
            )
             roleRef=dict(
                   kind= Role,
                   name= team_name,
                   apiGroup=rbac.authorization.k8s.io,
                    )
              )

         with open(role_filename, 'w') as outfile:
                  yaml.dump(yml_file_kubernetes_data, outfile, default_flow_style=False)

        subprocess.check_call(['kubectl', 'create', 'rolebinding', rolebinding_filename, '--namespace'+'='+team_name])

This isn't completely finished, but I want to make sure I am not doing it all wrong.

Comment (oliverdain) [reply new_app_setup.py]: that looks basically right. As per other comments I think `{}` is more common/Python-ic than `dict(...)` but it doesn't really matter much.

Also, isn't this:

```
'add'+'-'+'iam'+'-'+'policy'+'-'+'binding'
```

a long way to say this:

```
'add-iam-policy-binding'
```

File: new_app_setup.py (hunk 18 above)
Comment (oliverdain) [reply new_app_setup.py]: yes, that looks right.

File: new_app_setup.py (hunk 19 above)
Comment (oliverdain) [reply new_app_setup.py]: I think that's right.

File: new_app_setup.py (hunk 6 above)
Comment (alyssakelley) [reply new_app_setup.py]: I am still having issues with adding the users, but since we are focuing on the namespace and resourcequota for this PR, I will include that in a new branch.

File: new_app_setup.py [hunk 22]
```diff
@@ -22,64 +17,52 @@ def main():
 
     parser = argparse.ArgumentParser(description='Usage: python3 '
                                      'new_app_setup.py <Team name> '
-                                     '<List of users>')
+                                     '<Group user> '
+                                     '<Optional - List of users> '
+                                     '<Optional - Requested RAM> '
+                                     '<Optional - Requested CPU>')
```
Comment (oliverdain) [inline new_app_setup.py:23]: Since things like the CPU limits are arguments with flags like `--CPU` the help will automatically print info about each argument. If you document them as you have here people will think they're positional argument. The common usage message would be something like:

```
usage: python3 new_app_setup.py [options] 
```

and then the options docs are auto-generated by argparse.

File: new_app_setup.py [hunk 23]
```diff
@@ -22,64 +17,52 @@ def main():
...
                         'will be used for the entire group as a whole.')
 
     parser.add_argument('-LOU', '--list_of_users',
                         required=False,
                         help='A list of the users for this team.')
 
     parser.add_argument('-RAM', '--RAM', type=int, required=False,
-                        help='How many gigabytes of RAM your team is '
-                        'requesting, the deault is 2G.')
+                        default=2, help="How many gigabytes of RAM your "
```
Comment (oliverdain) [inline new_app_setup.py:37]: linux/unix standard is that options preceeded by a single `-` have only a single letter and options preceded by a double `--` can have multiple letters. You don't need to have the single `-` variant but it's nice for common options. So, for the RAM options it'd typically be `add_argument('-r', '--RAM', ...)` so that `-r 1` and `--RAM=10` both mean the same thing. If you don't want the single letter version of the flag then `add_argument('--RAM', ...)` is fine.

File: new_app_setup.py [hunk 24]
```diff
@@ -22,64 +17,52 @@ def main():
...
+    kube_apply(namespace_filename, team_name, "namespace")
 
-    create_kubernetes(namespace_filename, team_name, "namespace")
+    quota_filename = team_name + "_quota.yml"
 
-    if args.RAM is not None:
-        ram_requested = args.RAM
-    else:
-        ram_requested = '2G'  # default
+    create_ResourceQuota(quota_filename, team_name, args.RAM, args.CPU)
```
Comment (oliverdain) [inline new_app_setup.py:56]: method naming for pep8 would be `create_resource_quota`.

File: new_app_setup.py [hunk 25]
```diff
@@ -104,163 +87,114 @@ def create_kubernetes_namespace_yml(namespace_filename, team_name):
...
+        {
+            'name': team_name,
+        },
+        'spec':
+        {
+            'hard':
+            {
+                'requests.cpu': str(CPU),
+                'requests.memory': (str(RAM)+"G"),
+                'limits.cpu': "1",
```
Comment (oliverdain) [inline new_app_setup.py:106]: should come from the `CPU` value passed in - you have it hard coded to 1 here.

File: new_app_setup.py [hunk 26]
```diff
@@ -104,163 +87,114 @@ def create_kubernetes_namespace_yml(namespace_filename, team_name):
...
+                'limits.memory': "2G"
```
Comment (oliverdain) [inline new_app_setup.py:107]: as above, this should come from the value that was passed in.

Basically a pod can *request* a certain amount of RAM or CPU. It won't get scheduled anywhere unless k8's can guarantee it that amount of RAM or CPU. So your `request.cpu` and `request.memory` lines above limit how much a pod can *request*. Suppose a pod requests 1 CPU but gets scheduled on a machine with 4 CPU and it's the only thing scheduled there. By default, kubernetes will let it use all the available resources until something else gets scheduled on the same node and requests some of those resources. However, a pod can also put a *limit* on how much CPU or RAM it'll use regardless or what's available. Thats `limits.cpu` and `limits.memory`. Here you're saying the `request.cpu` is limited to the value of the CPU variable but you're also saying that the *limit* is 1. So if we had a team and we set their CPU to 2 it wouldn't work because you've said that no matter what they can't actually use more than 1 CPU.

You can read more about it here: https://kubernetes.io/docs/tasks/configure-pod-container/assign-cpu-resource/#motivation-for-cpu-requests-and-limits.

In any event, here you want *either*:

1. The `requests` and `limits` match: we limit each team to a fixed amount even if more is available
2. You specify `requests` but *not* `limits`: each team can request/be guaranteed more than a certain limit but if there's more resources available they can use them.

I suggest option (1). Why? Because if a team thinks they're running under their request limit but they actually require more we want them to know that before some other pod gets scheduled and causes them to get killed.

File: new_app_setup.py [hunk 27]
```diff
@@ -104,163 +87,114 @@ def create_kubernetes_namespace_yml(namespace_filename, team_name):
...
+    get_congif_yml_file and the parsing it to store only the access
+    token in a separate file so we can hand that to the Hack-For-A-
+    Cause teams. The file name for the tokens would be
+    access_token_from_config_<team_name>.txt and is placed in the
+    current directory with the other yml files created.
+    """
+    with open(config_file, "r") as file_contents:
+        dict_of_contents = yaml.load(file_contents)
+
+    filename_for_token = "access_token_from_config_"+team_name+".txt"
```
Comment (oliverdain) [inline new_app_setup.py:172]: run pep8 on this. I think it requires spaces around the `+` characters

File: new_app_setup.py [hunk 28]
```diff
@@ -104,163 +87,114 @@ def create_kubernetes_namespace_yml(namespace_filename, team_name):
...
+
+    access_token = (dict_of_contents['users'][0]['user']['auth-provider']
```
Comment (oliverdain) [inline new_app_setup.py:174]: I think this is just pulling out the access token for the 1st user in the entire cluster which is me. What we need to do is create a new user for the team and get the kubeconfig file for that user.

There's several ways to do this: https://cloud.google.com/iam/docs/overview. Scott and I are meeting today and will discuss options. In any event this needs to do the following:

1. Add a user to the cluster: https://cloud.google.com/kubernetes-engine/docs/how-to/iam. This will be done via `gcloud` commands
2. Add RBAC policies to limit that user to just their namespace: https://cloud.google.com/kubernetes-engine/docs/how-to/role-based-access-control. This will be done via `kubectl` commands.
3. Get a kubeconfig file for the team probably via `gcloud container clusters get-credentials ` (unless we go with service accounts)

This is all a bit complex so I might take this PR and add that stuff to it later today.

**Review by oliverdain — CHANGES_REQUESTED** [review]
Coming along nicely!

Sorry it took me so long to get to this.
